build_system/src/test.rs

```diff
@@ -876,54 +881,90 @@ where
     }
 
     if test_type == "ui" {
-        walk_dir(
-            rust_path.join("tests/ui"),
-            |dir| {
-                let dir_name = dir.file_name().and_then(|name| name.to_str()).unwrap_or("");
-                if [
-                    "abi",
-                    "extern",
-                    "unsized-locals",
-                    "proc-macro",
-                    "threads-sendsync",
-                    "borrowck",
-                    "test-attrs",
-                ]
-                .iter()
-                .any(|name| *name == dir_name)
-                {
-                    std::fs::remove_dir_all(dir).map_err(|error| {
-                        format!("Failed to remove folder `{}`: {:?}", dir.display(), error)
-                    })?;
+        if let Some(callback) = should_run_test_callback {
+            fn walk_dir<F, G>(
```

Comment:
Why is this new function necessary? Can't you reuse the existing `walk_dir`?

Reply:
Normal walk_dir doesn't allow walking through subdirectories I guess, thus had to implement this.

Reply:
Would it work to add a parameter `recursive` to the existing function instead?

Reply:
Ah! not sure how? I am aware of one walkdir crate that does it. Could you tell me if you if you know a way here. Moreover, custom functions are defined [here](https://github.com/lordshashank/rustc_codegen_gcc/blob/ad82635689a67aa0834020ac2f166a4c4ad6bb2e/build_system/src/test.rs#L957-L962) as well, though some customization was needed. Let me know if you have a way here.

Reply:
I believe it should be enough that to wrap the recursive call to `walk_dir()` with an `if recursive { […] }`.

Reply:
@antoyo I think I am not able to understand what you mean here, is it something like this
```
fn walk_dir<F, G>(
    dir_path: PathBuf,
    dir_callback: F,
    file_callback: G,
    recursive: bool,
) -> Result<(), String>
where
    F: Fn(&Path) -> Result<(), String> + Copy,
    G: Fn(&Path) -> Result<(), String> + Copy,
{
    if dir_path.is_dir() {
        for entry in std::fs::read_dir(dir_path).unwrap() {
            let entry = entry;
            let path = entry.unwrap().path();
            if path.is_dir() {
                dir_callback(&path)?;
                if recursive {
                    walk_dir(path, dir_callback, file_callback, recursive)?; // Recursive call
                }
            } else if path.is_file() {
                file_callback(&path)?;
            }
        }
    }
    Ok(())
}

walk_dir(
    rust_path.join("tests/ui"),
    |_dir| Ok(()),
    |file_path| {
        if contains_ui_error_patterns(file_path)? {
            Ok(())
        } else {
            remove_file(file_path).map_err(|e| e.to_string())
        }
    },
    true, // Set recursive to true
)?;
```

Reply:
Yes, but please change the `walk_dir` function in the file `build_system/src/utils.rs` since it correctly handles the errors.

Reply:
@antoyo did that but am getting recursion limit error 
```
error: reached the recursion limit while instantiating `walk_dir::<PathBuf, &mut &mut &mut &mut ..., ...>`
   --> src/utils.rs:362:17
    |
362 |                 walk_dir(entry_path, &mut dir_cb, &mut file_cb, recursive)?; // Recursive call
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: `walk_dir` defined here
   --> src/utils.rs:346:1
    |
    ```
    would have to use `Box<dyn FnMut>` to avoid this. what to do?

Reply:
Can you please post the complete code of the function?
Also, does it work if you do:
```
walk_dir(entry_path, dir_cb, file_cb, recursive)?; // Recursive call
```

Reply:
here is the function
```

pub fn walk_dir<P, D, F>(dir: P, mut dir_cb: D, mut file_cb: F, recursive: bool) -> Result<(), String>
where
    P: AsRef<Path>,
    D: FnMut(&Path) -> Result<(), String>,
    F: FnMut(&Path) -> Result<(), String>,
{
    let dir = dir.as_ref();
    for entry in fs::read_dir(dir)
        .map_err(|error| format!("Failed to read dir `{}`: {:?}", dir.display(), error))?
    {
        let entry = entry
            .map_err(|error| format!("Failed to read entry in `{}`: {:?}", dir.display(), error))?;
        let entry_path = entry.path();
        if entry_path.is_dir() {
            dir_cb(&entry_path)?;
            if recursive {
                walk_dir(entry_path, &mut dir_cb, &mut file_cb, recursive)?; // Recursive call
            }
        } else {
            file_cb(&entry_path)?;
        }
    }
    Ok(())
}
```
Also @antoyo, removing `&mut` would lead to borrowing of moved value, hence used that

Reply:
I guess this would work:
```
pub fn walk_dir<P, D, F>(dir: P, dir_cb: &mut D, file_cb: &mut F, recursive: bool) -> Result<(), String>
where
    P: AsRef<Path>,
    D: FnMut(&Path) -> Result<(), String>,
    F: FnMut(&Path) -> Result<(), String>,
{
    let dir = dir.as_ref();
    for entry in fs::read_dir(dir)
        .map_err(|error| format!("Failed to read dir `{}`: {:?}", dir.display(), error))?
    {
        let entry = entry
            .map_err(|error| format!("Failed to read entry in `{}`: {:?}", dir.display(), error))?;
        let entry_path = entry.path();
        if entry_path.is_dir() {
            dir_cb(&entry_path)?;
            if recursive {
                walk_dir(entry_path, dir_cb, file_cb, recursive)?; // Recursive call
            }
        } else {
            file_cb(&entry_path)?;
        }
    }
    Ok(())
}
```
This way, you will need to add `&mut` at the call-site for the callback arguments.